### PR TITLE
Network improvements

### PR DESCRIPTION
**Pull request by szok**
- Add ordering in IPAddress and Network page
- Add ordering to network columns
- Add update network parents after save

### CODE REVIEW

File: src/ralph/networks/models/networks.py [hunk 1]
```diff
@@ -242,6 +242,16 @@ def save(self, *args, **kwargs):
...
         self._assign_ips_to_network()
+        self._update_subnetworks_parent()
+
+    def _update_subnetworks_parent(self):
+        parent = self.get_network()
+        networks = self.__class__.objects.filter(
+            parent=parent
+        ).exclude(pk=self.pk)
+        for network in networks:
+            network.parent = network.get_network()
```
Comment (mkurek) [inline src/ralph/networks/models/networks.py]: `save` already take care of this

File: src/ralph/networks/models/networks.py [hunk 2]
```diff
@@ -242,6 +242,16 @@ def save(self, *args, **kwargs):
         self._assign_parent()
         super(Network, self).save(*args, **kwargs)
         self._assign_ips_to_network()
+        self._update_subnetworks_parent()
+
+    def _update_subnetworks_parent(self):
```
Comment (mkurek) [inline src/ralph/networks/models/networks.py:275]: i'd appreciate some tests for it

File: src/ralph/networks/admin.py [hunk 3]
```diff
@@ -162,6 +168,16 @@ def address(self, obj):
     address.short_description = _('Network address')
     address.admin_order_field = ['min_ip']
```
Comment (mkurek) [inline src/ralph/networks/admin.py]: min and max

File: src/ralph/networks/admin.py [hunk 4]
```diff
@@ -208,6 +224,12 @@ def show_addresses(self, network):
     show_addresses.allow_tags = True
     show_addresses.short_description = _('Addresses')
 
+    def get_queryset(self, request):
+        qs = super().get_queryset(request)
+        qs = qs.annotate(ipaddress_count=Count('ips'))
+        qs = qs.extra(select={'subnetwork_count': 'rght - lft'})
```
Comment (mkurek) [inline src/ralph/networks/admin.py]: add a comment that this is not exact number of subnetworks but it's fine estimate for ordering

Comment (mkurek) [inline src/ralph/networks/admin.py]: btw you could paste a link to the mptt here (https://github.com/django-mptt/django-mptt/blob/master/mptt/models.py#L594)

File: src/ralph/networks/admin.py [hunk 5]
```diff
@@ -270,7 +293,7 @@ def get_queryset(self, request):
         )
 
     def ip_address(self, obj):
```
Comment (mkurek) [inline src/ralph/networks/admin.py:297]: do we still need this function? if you'll remove it, remember that sorting by ip should be sorting by number underneath

File: src/ralph/networks/tests/test_models.py [hunk 6]
```diff
@@ -197,6 +197,25 @@ def test_sub_network_should_assign_automatically(self):
...
+
+        net3 = Network.objects.create(
+            name='net3', address='10.20.30.128/25'
+        )
+
+        net2 = Network.objects.get(pk=net2.pk)
+        self.assertEqual(net2.parent, net3)
+
+        self.assertEqual(net3.parent, net1)
+
```
Comment (mkurek) [inline src/ralph/networks/tests/test_models.py:218]: additional tests:
- add network `10.20.40.128/28` as net4
- change net3 address now to `10.20.40.128/25`
- now net4 should be child of net3, and net2 should be child of net1
- remove net3 - now net4 should be parentless

File: src/ralph/networks/admin.py [hunk 7]
```diff
@@ -162,6 +168,16 @@ def address(self, obj):
     address.short_description = _('Network address')
     address.admin_order_field = ['min_ip']
 
+    def subnetwork_count(self, obj):
```
Comment (mkurek) [inline src/ralph/networks/admin.py]: subnetworks_count

File: src/ralph/networks/admin.py (hunk 5 above)
Comment (szok) [inline src/ralph/networks/admin.py:297]: Yes, only for column ordering.

Comment (mkurek) [inline src/ralph/networks/admin.py:297]: i really don't like wrapping it in `<a ...>` - what about this (on model):

```
@property
def _address_sortable(self):
    return self.address
_address_sortable.admin_order_field = 'number'
```

File: src/ralph/networks/models/networks.py [hunk 8]
```diff
@@ -82,9 +82,7 @@ class NetworkMixin(object):
     _parent_attr = None
 
     def _assign_parent(self):
-        parent = getattr(self, self._parent_attr)
-        if not parent or self.pk:
-            setattr(self, self._parent_attr, self.get_network())
+        setattr(self, self._parent_attr, self.get_network())
```
Comment (mkurek) [inline src/ralph/networks/models/networks.py:88]: why?

**Comment by mkurek** [comment]
:ok:

File: src/ralph/networks/tests/test_models.py [hunk 9]
```diff
@@ -197,6 +197,39 @@ def test_sub_network_should_assign_automatically(self):
...
+        net3 = Network.objects.create(
+            name='net3', address='10.20.30.128/25'
+        )
+
+        self.refresh_objects_from_db(net2)
+        self.assertEqual(net2.parent, net3)
+
+        self.assertEqual(net3.parent, net1)
+
+        net4 = Network.objects.create(
```
Comment (mkurek) [inline src/ralph/networks/tests/test_models.py:219]: please add here `net5` with address `10.20.40.0/24` and check if net3 after changing address will be descendant of net5

File: src/ralph/networks/tests/test_models.py [hunk 10]
```diff
@@ -197,6 +197,39 @@ def test_sub_network_should_assign_automatically(self):
...
+        net3.address = '10.20.40.128/25'
+        net3.save()
+
+        self.refresh_objects_from_db(net1, net2, net3, net4)
+
+        self.assertEqual(net4.parent, net3)
+        self.assertEqual(net2.parent, net1)
+
+        net3.delete()
+        self.assertIsNone(Network.objects.filter(pk=net4.pk).first())
```
Comment (mkurek) [inline src/ralph/networks/tests/test_models.py]: is this desired behavior that deleting parent network will delete all subnetworks as well? imo `net4` should stay as it were before, but without parent now

**Comment by ar4s** [comment]
:ok:

File: src/ralph/networks/tests/test_models.py (hunk 9 above)
Comment (szok) [inline src/ralph/networks/tests/test_models.py:219]: Done

File: src/ralph/networks/models/networks.py [hunk 11]
```diff
@@ -257,11 +255,24 @@ def save(self, *args, **kwargs):
...
         self.max_ip = int(self.broadcast_address)
         self._assign_parent()
         super(Network, self).save(*args, **kwargs)
         self._assign_ips_to_network()
+        if enable_save_descendants:
+            self._update_subnetworks_parent()
+
+    def delete(self):
+        with transaction.atomic():
+            self.address = '0.0.0.0/32'
```
Comment (mkurek) [inline src/ralph/networks/models/networks.py:271]: add some comment why this fake address is used

**Pull request merged**